Fix issue product name in minicart render wrong with special characters

mrtuvn opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
With product has url magento render name without problem. But in the case product without url will render product name in cart incorrectly. Ex: product with special characters like double quotes
This fix will cover the case products don't have url and render name correctly


### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#29075

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. From backend admin create new product with name contain special character such as double quote, &, -, /
2. Clear cache go to frontend view new product

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->
I have one question need clarify about knockout 
In template inside foreach loop when use $parent for reference to component i can't call to method that i have defined But if i used $parents[1] its work to call method! Any reason behind this ? i don't see any doc mention in magento official devdocs
Is that knockout context changed inside loop foreach ?
CC: @ihor-sviziev @omiroshnichenko @guz-anton

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

ihor-sviziev commented:
@magento run all tests

ihor-sviziev reviewed (changes requested):
@mrtuvn could you fix the failing static tests?

mrtuvn commented:
@ihor-sviziev just find out this rule in validate code magento
https://devdocs.magento.com/guides/v2.4/extension-dev-guide/xss-protection.html

> In order to notify developers that these properties/function results may contain HTML, Magento requires (with the help of a static test) that you name such properties/functions using “UnsanitizedHtml” suffix.

mrtuvn commented:
Need your opinions ! @lbajsarowicz @lenaorobei @sivaschenko

ihor-sviziev commented:
@mrtuvn I believe we need to ignore these cases as we have html there for sure

mrtuvn commented:
What should i do on this ? Not only the product name, still other place in this file fail static test 
https://github.com/magento/magento2/blob/32b520c3b34b2da7dc253ee1dd448cae4b072bc8/app/code/Magento/Checkout/view/frontend/web/template/minicart/item/default.html#L48
This line will violation static rule magento about html sanitize

gabrieldagama commented:
The risk was set to`low` due to: the suggested changes shouldn't affect other areas.

engcom-Charlie commented:
@mrtuvn please see https://devdocs.magento.com/guides/v2.4/extension-dev-guide/xss-protection.html about `Knockout templates` this should help you solve the problem.

ihor-sviziev reviewed (changes requested):
@mrtuvn from what @engcom-Charlie shown above - seems like we need to change the variable name from `product_name` to `productNameUnsanitizedHtml`, revert change that was done in https://github.com/magento/magento2/pull/13802/files#diff-6502b94b418435d6341605ae8a37b46fR88 and add new variable `productNameUnsanitizedHtml` instead.
Could you do that?

mrtuvn commented:
yes i can do that but seem need more effort that still have other variable like options
https://github.com/afirlejczyk/magento2/blob/764f0862f46204f649367728333f40f39509bd7c/app/code/Magento/Checkout/view/frontend/web/template/minicart/item/default.html#L45

this value seem previously passed validator check

ihor-sviziev commented:
@mrtuvn can we move it to some method?  I believe if that method will be named correctly - it should not complain

mrtuvn commented:
@ihor-sviziev 
I not sure how to pass static test in the 2 last line
https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Checkout/view/frontend/web/template/minicart/item/default.html#L45
and Line 48

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run all tests

ihor-sviziev commented on app/code/Magento/Checkout/CustomerData/DefaultItem.php:
```diff
@@ -92,7 +92,7 @@ protected function doGetItemData()
             'is_visible_in_site_visibility' => $this->item->getProduct()->isVisibleInSiteVisibility(),
             'product_id' => $this->item->getProduct()->getId(),
-            'product_name' => $productName,
+            'productNameUnsanitizedHtml' => $productName,
```
I think we need to bring back product_name, probably would be better to have it not escaped. What do you thing ?

mrtuvn replied:
yes i will revert that

mrtuvn commented:
@nuzil @sivaschenko @sidolov
 Sorry for involved you guys here ! Can you guys give me some suggestion about my changes

ihor-sviziev commented on app/code/Magento/Checkout/view/frontend/web/js/view/cart-item-renderer.js:
```diff
@@ -0,0 +1,42 @@
+         * @return {String}
+         */
+        getProductNameUnsanitizedHtml: function (productName) {
+            return escaper.escapeHtml(productName, this.allowedTags);
```
Looks like now we have 2 times escaping, am I correct? Does it works fine now?

mrtuvn replied:
just seperate function name for clarify imo. Both function do the same task for escape html string

ihor-sviziev replied:
I mean right now product name escaped on backend and on front end , isn’t it?

mrtuvn replied:
should we return product name ? Current in this PR we have 2 time escape in product name
One from backend and one in knockout template when we call getProductNameUnsanitizedHtml

ihor-sviziev replied:
I think it would be good, also please add comment that it’s escaped on backend

ihor-sviziev replied:
Also would it complain if we’ll use name `getProductNameHtml`? I think it shouldn’t and it is better name for the function

mrtuvn replied:
name method need have suffix UnsanitizedHtml otherwise static test will raise error
https://github.com/magento/magento2/blob/2.4-develop/dev/tests/static/framework/Magento/Sniffs/Html/HtmlBindingSniff.php#L79

ihor-sviziev replied:
Ok, so I think we can leave there something like this:
```suggestion
            // product name already escaped on backend
            return productName;
```

ihor-sviziev reviewed (commented):
Now your changes looks good to me. The only one thing is not clear to me - do we double escaping for product name now?

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented:
@magento run all tests

mrtuvn commented:
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

mrtuvn commented:
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

gabrieldagama commented:
@magento run all tests

ihor-sviziev commented:
Now changes looks good to me. Let's wait for test results. 

@magento run all tests

mrtuvn commented:
Seem fail tests not related with changes

mrtuvn commented:
rebase with mainline latest code base

mrtuvn commented:
rebased branch with mainline latest code base

mrtuvn commented:
@magento run all tests

ihor-sviziev commented:
@magento give me test instance

ihor-sviziev commented:
Hi @mrtuvn i was able to reproduce the issue that we’re getting in families mftf tests:
1. Add to cart following product https://966e75ea4c12ca3b7972ea3f76a857c8.instances.magento-community.engineering/sprite-yoga-companion-kit.html
2. Open mini cart and see that we don’t have a price and here is some warning without message
![9CF0D6C9-2BEC-44CC-8717-22E0D5C57342](https://user-images.githubusercontent.com/1873745/93161979-68588c80-f71c-11ea-8320-a437afb73edc.png)

Could you take a look?

mrtuvn commented:
In my opinion current this time we don't need touch more on this cart for fix problem. This issue happen only by customises
But soon in near future we possible need refactor code and add some additional for pass new validator for html escape
Should i close this or continuing ?

ihor-sviziev commented on app/code/Magento/Checkout/view/frontend/web/js/view/cart-item-renderer.js:
```diff
@@ -0,0 +1,43 @@
+         * @return {String}
+         */
+        getOptionValueUnsanitizedHtml: function (optionValue) {
+            return escaper.escapeHtml(optionValue, this.allowedTags);
```
```suggestion
            // option value should be already escaped on backend
            return optionValue;
```

ihor-sviziev commented on app/code/Magento/Checkout/view/frontend/web/js/view/cart-item-renderer.js:
```diff
@@ -0,0 +1,43 @@
+    'use strict';
+
+    return Component.extend({
+        defaults: {
+            allowedTags: ['div', 'span', 'b', 'strong', 'i', 'em', 'u', 'a']
+        },
+
+        /** @inheritdoc */
+        initialize: function () {
+            this._super();
+        },
+
```
I think we can remove it
```suggestion
```

ihor-sviziev reviewed (changes requested):
I just analyzed this HTML one time ago - seems like option.value in cases that you changed should contain already escaped html, so it should come from backend.
I would suggest just to remove escaping on frontend and add comment that it should be already escaped on backend.

As result we'll just fix the issue with product_name and will not change anything related to option.value.

What do you thing?

mrtuvn committed:
Fix issue product name render wrong with special characters


update


update


update


update


up


up


up


update


up


Update default.html

fix getProductNameUnsanitizedHtml is not defined
up

ihor-sviziev commented:
@magento run all tests

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8215](https://jira.corp.magento.com/browse/ENGCOM-8215) has been created to process this Pull Request

engcom-Alfa commented:
:heavy_check_mark: QA Passed

**Manual testing scenario:**

1. Add product with special characters to cart
2. remove ifnot condition (or trigger usage of it in other way) in template https://github.com/magento/magento2/blob/024625e61887a252df56f027e42347d20860b253/app/code/Magento/Checkout/view/frontend/web/template/minicart/item/default.html#L24-L32
3. open minicart

**Before:** :heavy_multiplication_x: incorrectly renderer product name in the cart 

![Screenshot from 2020-09-21 12-59-50](https://user-images.githubusercontent.com/51679138/93756520-e7166380-fc0d-11ea-8c8f-a69b5ebdf488.png)

**After:** :heavy_check_mark: correctly rendered product name in the cart

![Screenshot from 2020-09-21 13-10-27](https://user-images.githubusercontent.com/51679138/93756606-04e3c880-fc0e-11ea-91fc-eef7a6eb2fee.png)
